<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Jun Zhang">
<title>Global gene expression analysis – RiboTransVis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Pathway-enrichment-analysis.html" rel="next">
<link href="./Track-plot.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-f1075ac1703e165f0c9b3df650605ffb.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script defer="" data-domain="r4ds.hadley.nz" src="https://plausible.io/js/plausible.js"></script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Gene-expression-analysis.html">Global gene expression analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">RiboTransVis</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/junjunlab/riboTransVis" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./Preprocessing-into.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preprocessing introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Preprocessing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Preprocessing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Extract-sequence.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Extract sequence</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Object-construct.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Object construct</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./Quality-assessment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality assessment</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Ribo-QC.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Ribo-seq QC check</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Length-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Length distribution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Frame-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">In frame ratio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Periodicy-check.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Periodicity check</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Feature-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Feature distribution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Meta-frame-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Meta frame plot</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Meta-heatmap-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Meta heatmap plot</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Offset-check.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Offset check</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Whole-metagene-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Whole metegene plot</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Metagene-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Metagene plot</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Offset-correction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Offset correction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Polarity-calculation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polarity score</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Codon-occupancy.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Codon occupancy</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Multipeptide-occupancy.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multiple peptides occupancy</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Peptide-motif-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Peptide motif plot</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Motif-centered.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Motif centered analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Track-plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Track plot</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Gene-expression-analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Global gene expression analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Pathway-enrichment-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pathway enrichment analysis</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./More-examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More examples</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Ubiquitin-conjugase-Rad6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Ubiquitin conjugase Rad6</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./DENR-promotes-translation-reinitiation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">DENR promotes translation reinitiation</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#intro" id="toc-intro" class="nav-link active" data-scroll-target="#intro">Intro</a></li>
  <li><a href="#quantify-gene-expression-with-counts" id="toc-quantify-gene-expression-with-counts" class="nav-link" data-scroll-target="#quantify-gene-expression-with-counts">Quantify gene expression with counts</a></li>
  <li><a href="#correlation-plot" id="toc-correlation-plot" class="nav-link" data-scroll-target="#correlation-plot">Correlation plot</a></li>
  <li><a href="#extract-normalized-expression" id="toc-extract-normalized-expression" class="nav-link" data-scroll-target="#extract-normalized-expression">Extract normalized expression</a></li>
  <li><a href="#differential-analysis" id="toc-differential-analysis" class="nav-link" data-scroll-target="#differential-analysis">Differential analysis</a></li>
  <li><a href="#translation-efficiency-differential-analysis" id="toc-translation-efficiency-differential-analysis" class="nav-link" data-scroll-target="#translation-efficiency-differential-analysis">Translation efficiency differential analysis</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/junjunlab/riboTransVis/edit/main/Gene-expression-analysis.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/junjunlab/riboTransVis/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Global gene expression analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="intro" class="level2"><h2 class="anchored" data-anchor-id="intro">Intro</h2>
<p>In addition to analyzing ribosome profiling (Ribo-seq) data from the perspective of codon usage, amino acid composition, or other sequence-based features, a key focus in translational research lies in assessing changes in gene-level expression and translational regulation. Specifically, researchers are often interested in how translational efficiency (TE) and mRNA abundance vary across different conditions.</p>
<p>This section introduces how to use the riboTransVis toolkit for global RNA and RPF (Ribosome-Protected Fragment) quantification, differential expression analysis, and calculation of translational efficiency. By integrating Ribo-seq and RNA-seq data, riboTransVis facilitates systematic identification of genes with altered protein synthesis, enabling insights into translational control mechanisms that cannot be captured by transcriptome data alone.</p>
<p>Through this approach, we aim to provide a comprehensive framework for analyzing genome-wide translational changes, which can support functional discovery and reveal phenomena such as selective translation, stress response adaptation, and oncogenic translation reprogramming.</p>
</section><section id="quantify-gene-expression-with-counts" class="level2"><h2 class="anchored" data-anchor-id="quantify-gene-expression-with-counts">Quantify gene expression with counts</h2>
<p>The <code>get_counts()</code> function utilizes the <strong>featureCounts</strong> algorithm from the Rsubread package to efficiently quantify gene-level expression from RNA-seq and Ribo-seq data. By default, it performs read counting over exon features for RNA-seq and coding sequence (<strong>CDS</strong>) features for RPF (ribosome-protected fragments) data. Users can explicitly specify the annotation features used for quantification via the options: <strong>rna_feature = “exon”</strong> and <strong>ribo_feature = “CDS”</strong>.</p>
<p>Additionally, if the BAM files are aligned to the genome, Rsubread’s featureCounts is used for quantification. If instead the BAM files are aligned to the transcriptome, the function directly obtains read counts from the BAM files themselves.</p>
<p>The resulting count information is stored in the <strong>counts</strong> slot of the input object:</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">get_counts</span><span class="op">(</span><span class="va">obj</span>, nThreads <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co"># List of 2</span></span>
<span><span class="co"># $ rpf:'data.frame':   14647 obs. of  4 variables:</span></span>
<span><span class="co">#   ..$ wt1: num [1:14647] 1 15 482 616 239 935 262 101 647 274 ...</span></span>
<span><span class="co"># ..$ wt2: num [1:14647] 1 15 482 616 239 935 262 101 647 274 ...</span></span>
<span><span class="co"># ..$ ko1: num [1:14647] 0 43 491 783 323 ...</span></span>
<span><span class="co"># ..$ ko2: num [1:14647] 0 21 280 464 90 713 232 64 543 290 ...</span></span>
<span><span class="co"># $ rna:'data.frame':   57169 obs. of  4 variables:</span></span>
<span><span class="co">#   ..$ wt1: num [1:57169] 6 1 2 0 0 2 5 7 0 0 ...</span></span>
<span><span class="co"># ..$ wt2: num [1:57169] 6 1 2 0 0 2 5 7 0 0 ...</span></span>
<span><span class="co"># ..$ ko1: num [1:57169] 6 0 1 0 0 1 6 9 0 0 ...</span></span>
<span><span class="co"># ..$ ko2: num [1:57169] 3 1 0 0 0 1 7 1 0 0 ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="correlation-plot" class="level2"><h2 class="anchored" data-anchor-id="correlation-plot">Correlation plot</h2>
<p>Plot the correlation for the replicates based on counts data:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">counts</span><span class="op">$</span><span class="va">rna</span></span>
<span></span>
<span><span class="fu">cor_plot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,x <span class="op">=</span> <span class="st">"ko1"</span>,y <span class="op">=</span> <span class="st">"ko2"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="image/corp.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</section><section id="extract-normalized-expression" class="level2"><h2 class="anchored" data-anchor-id="extract-normalized-expression">Extract normalized expression</h2>
<p>The function <code>get_normalized_reads</code> calculates <strong>RNA</strong>, <strong>RPF</strong>, and <strong>translation efficiency (TE)</strong>. The normalized values for RNA and RPF are standardized using TPM values, while translation efficiency (TE) is defined as:</p>
<p><span class="math display">\[
\text{TE (translation efficiency)} = \frac{\text{RPF}_tpm}{\text{RNA}_tpm} \tag{1}
\]</span></p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">norm</span> <span class="op">&lt;-</span> <span class="fu">get_normalized_reads</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>                             type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"ribo"</span>, <span class="st">"te"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">norm</span><span class="op">)</span></span>
<span><span class="co"># List of 3</span></span>
<span><span class="co"># $ tpm.rna :'data.frame':  57169 obs. of  5 variables:</span></span>
<span><span class="co">#   ..$ wt1      : num [1:57169] 0.2551 0.0344 0.0154 0 0 ...</span></span>
<span><span class="co"># ..$ wt2      : num [1:57169] 0.2551 0.0344 0.0154 0 0 ...</span></span>
<span><span class="co"># ..$ ko1      : num [1:57169] 0.22734 0 0.00686 0 0 ...</span></span>
<span><span class="co"># ..$ ko2      : num [1:57169] 0.1312 0.0354 0 0 0 ...</span></span>
<span><span class="co"># ..$ gene_name: chr [1:57169] "A1BG" "A1BG-AS1" "A1CF" "A2M" ...</span></span>
<span><span class="co"># $ tpm.ribo:'data.frame':  14647 obs. of  5 variables:</span></span>
<span><span class="co">#   ..$ wt1      : num [1:14647] 0.0257 0.6961 50.8271 81.9381 17.6403 ...</span></span>
<span><span class="co"># ..$ wt2      : num [1:14647] 0.0257 0.6961 50.8271 81.9381 17.6403 ...</span></span>
<span><span class="co"># ..$ ko1      : num [1:14647] 0 1.44 37.4 75.23 17.22 ...</span></span>
<span><span class="co"># ..$ ko2      : num [1:14647] 0 1.24 37.61 78.61 8.46 ...</span></span>
<span><span class="co"># ..$ gene_name: chr [1:14647] "A1CF" "A4GALT" "AAAS" "AACS" ...</span></span>
<span><span class="co"># $ te      :'data.frame':  14647 obs. of  5 variables:</span></span>
<span><span class="co">#   ..$ wt1      : num [1:14647] 1.672 0 0.394 8.632 19.743 ...</span></span>
<span><span class="co"># ..$ wt2      : num [1:14647] 1.672 0 0.394 8.632 19.743 ...</span></span>
<span><span class="co"># ..$ ko1      : num [1:14647] 0 9.397 0.427 9.333 22.229 ...</span></span>
<span><span class="co"># ..$ ko2      : num [1:14647] NaN 0 0.554 9.186 27.907 ...</span></span>
<span><span class="co"># ..$ gene_name: chr [1:14647] "A1CF" "A2ML1" "A4GALT" "AAAS" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-analysis" class="level2"><h2 class="anchored" data-anchor-id="differential-analysis">Differential analysis</h2>
<p>The <code>gene_differential_analysis</code> function uses <strong>DESeq2</strong> to perform differential expression analysis separately on RNA or RPF data and returns the corresponding results:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diff.rna</span> <span class="op">&lt;-</span> <span class="fu">gene_differential_analysis</span><span class="op">(</span><span class="va">obj</span>, </span>
<span>                                       type <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>                                       control_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"wt1"</span>,<span class="st">"wt2"</span><span class="op">)</span>,</span>
<span>                                       treat_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ko1"</span>,<span class="st">"ko2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">diff.rna</span><span class="op">)</span></span>
<span><span class="co">#            baseMean log2FoldChange    lfcSE       stat    pvalue padj gene_name   type</span></span>
<span><span class="co"># A1BG      5.0888259     -0.2345798 1.248392 -0.1879055 0.8509508   NA      A1BG nonSig</span></span>
<span><span class="co"># A1BG-AS1  0.7829185     -0.6432696 3.213131 -0.2002002 0.8413240   NA  A1BG-AS1 nonSig</span></span>
<span><span class="co"># A1CF      1.1471459     -1.8219647 2.754492 -0.6614522 0.5083223   NA      A1CF nonSig</span></span>
<span><span class="co"># A2ML1     1.4697527     -0.7623221 2.236789 -0.3408109 0.7332459   NA     A2ML1 nonSig</span></span>
<span><span class="co"># A2ML1-AS1 5.9189416      0.6372937 1.177298  0.5413188 0.5882879   NA A2ML1-AS1 nonSig</span></span>
<span><span class="co"># A2ML1-AS2 5.5834917     -0.4017920 1.313643 -0.3058609 0.7597106   NA A2ML1-AS2 nonSig</span></span>
<span></span>
<span><span class="va">diff.ribo</span> <span class="op">&lt;-</span> <span class="fu">gene_differential_analysis</span><span class="op">(</span><span class="va">obj</span>, </span>
<span>                                        type <span class="op">=</span> <span class="st">"ribo"</span>,</span>
<span>                                        control_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"wt1"</span>,<span class="st">"wt2"</span><span class="op">)</span>,</span>
<span>                                        treat_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ko1"</span>,<span class="st">"ko2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">diff.ribo</span><span class="op">)</span></span>
<span><span class="co">#           baseMean log2FoldChange     lfcSE       stat      pvalue      padj gene_name   type</span></span>
<span><span class="co"># A1CF     0.5143687     -2.5281257 4.5894024 -0.5508616 0.581728523        NA      A1CF nonSig</span></span>
<span><span class="co"># A4GALT  22.3216385      0.9254935 0.6865945  1.3479477 0.177675211 0.6322686    A4GALT nonSig</span></span>
<span><span class="co"># AAAS   427.6743989     -0.4642721 0.1766959 -2.6275206 0.008600965 0.1247037      AAAS nonSig</span></span>
<span><span class="co"># AACS   609.1352834     -0.1178472 0.1558609 -0.7561051 0.449586198 0.8723837      AACS nonSig</span></span>
<span><span class="co"># AADAT  210.8025653     -0.4692803 0.3222716 -1.4561639 0.145347309 0.5805468     AADAT nonSig</span></span>
<span><span class="co"># AAGAB  921.2909124     -0.1293514 0.1392711 -0.9287745 0.353005973 0.8165758     AAGAB nonSig</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="translation-efficiency-differential-analysis" class="level2"><h2 class="anchored" data-anchor-id="translation-efficiency-differential-analysis">Translation efficiency differential analysis</h2>
<p>The <code>TE_differential_analysis</code> function performs a differential analysis of translation efficiency. Users can select either <a href="https://github.com/smithlabcode/riborex"><strong>RiboReX</strong></a> or <a href="https://github.com/xryanglab/xtail"><strong>Xtail</strong></a> as their preferred tool for this analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># use riborex for te analysis</span></span>
<span><span class="va">te.rex</span> <span class="op">&lt;-</span> <span class="fu">TE_differential_analysis</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>                                   pkg <span class="op">=</span> <span class="st">"riborex"</span>,</span>
<span>                                   control_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"wt1"</span>,<span class="st">"wt2"</span><span class="op">)</span>,</span>
<span>                                   treat_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ko1"</span>,<span class="st">"ko2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">te.rex</span><span class="op">)</span></span>
<span><span class="co"># List of 2</span></span>
<span><span class="co"># $ deg_raw :Formal class 'DESeqResults' [package "DESeq2"] with 7 slots</span></span>
<span><span class="co"># .. ..@ priorInfo      : list()</span></span>
<span><span class="co"># .. ..@ rownames       : chr [1:13204] "A4GALT" "AAAS" "AACS" "AADAT" ...</span></span>
<span><span class="co"># .. ..@ nrows          : int 13204</span></span>
<span><span class="co"># .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co"># .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co"># .. .. .. ..@ rownames       : NULL</span></span>
<span><span class="co"># .. .. .. ..@ nrows          : int 6</span></span>
<span><span class="co"># .. .. .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co"># .. .. .. ..@ elementMetadata: NULL</span></span>
<span><span class="co"># .. .. .. ..@ metadata       : list()</span></span>
<span><span class="co"># .. .. .. ..@ listData       :List of 2</span></span>
<span><span class="co"># .. .. .. .. ..$ type       : chr [1:6] "intermediate" "results" "results" "results" ...</span></span>
<span><span class="co"># .. .. .. .. ..$ description: chr [1:6] "mean of normalized counts for all samples" "log2 fold change (MLE): EXTRA1 treated vs control" "standard error: EXTRA1 treated vs control" "Wald statistic: EXTRA1 treated vs control" ...</span></span>
<span><span class="co"># .. ..@ metadata       :List of 6</span></span>
<span><span class="co"># .. .. ..$ filterThreshold: Named num 93</span></span>
<span><span class="co"># .. .. .. ..- attr(*, "names")= chr "40.71429%"</span></span>
<span><span class="co"># .. .. ..$ filterTheta    : num 0.407</span></span>
<span><span class="co"># .. .. ..$ filterNumRej   :'data.frame':   50 obs. of  2 variables:</span></span>
<span><span class="co">#   .. .. .. ..$ theta : num [1:50] 0 0.0194 0.0388 0.0582 0.0776 ...</span></span>
<span><span class="co"># .. .. .. ..$ numRej: num [1:50] 112 116 116 117 118 118 118 120 121 124 ...</span></span>
<span><span class="co"># .. .. ..$ lo.fit         :List of 2</span></span>
<span><span class="co"># .. .. .. ..$ x: num [1:50] 0 0.0194 0.0388 0.0582 0.0776 ...</span></span>
<span><span class="co"># .. .. .. ..$ y: num [1:50] 114 115 116 117 117 ...</span></span>
<span><span class="co"># .. .. ..$ alpha          : num 0.1</span></span>
<span><span class="co"># .. .. ..$ lfcThreshold   : num 0</span></span>
<span><span class="co"># .. ..@ listData       :List of 6</span></span>
<span><span class="co"># .. .. ..$ baseMean      : num [1:13204] 56.7 246.1 341.4 143.7 568 ...</span></span>
<span><span class="co"># .. .. ..$ log2FoldChange: num [1:13204] 0.0106 -0.1939 0.0381 -0.1659 0.0415 ...</span></span>
<span><span class="co"># .. .. ..$ lfcSE         : num [1:13204] 0.578 0.285 0.241 0.54 0.208 ...</span></span>
<span><span class="co"># .. .. ..$ stat          : num [1:13204] 0.0183 -0.681 0.158 -0.3072 0.1996 ...</span></span>
<span><span class="co"># .. .. ..$ pvalue        : num [1:13204] 0.985 0.496 0.874 0.759 0.842 ...</span></span>
<span><span class="co"># .. .. ..$ padj          : num [1:13204] NA 0.915 0.983 0.968 0.98 ...</span></span>
<span><span class="co"># $ deg_anno:'data.frame':  13204 obs. of  8 variables:</span></span>
<span><span class="co">#   ..$ baseMean      : num [1:13204] 56.7 246.1 341.4 143.7 568 ...</span></span>
<span><span class="co"># ..$ log2FoldChange: num [1:13204] 0.0106 -0.1939 0.0381 -0.1659 0.0415 ...</span></span>
<span><span class="co"># ..$ lfcSE         : num [1:13204] 0.578 0.285 0.241 0.54 0.208 ...</span></span>
<span><span class="co"># ..$ stat          : num [1:13204] 0.0183 -0.681 0.158 -0.3072 0.1996 ...</span></span>
<span><span class="co"># ..$ pvalue        : num [1:13204] 0.985 0.496 0.874 0.759 0.842 ...</span></span>
<span><span class="co"># ..$ padj          : num [1:13204] NA 0.915 0.983 0.968 0.98 ...</span></span>
<span><span class="co"># ..$ gene_name     : chr [1:13204] "A4GALT" "AAAS" "AACS" "AADAT" ...</span></span>
<span><span class="co"># ..$ type          : chr [1:13204] "nonSig" "nonSig" "nonSig" "nonSig" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Xtail</strong> for differential analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">te.xtail</span> <span class="op">&lt;-</span> <span class="fu">TE_differential_analysis</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>                                     pkg <span class="op">=</span> <span class="st">"xtail"</span>,</span>
<span>                                     control_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"wt1"</span>,<span class="st">"wt2"</span><span class="op">)</span>,</span>
<span>                                     treat_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ko1"</span>,<span class="st">"ko2"</span><span class="op">)</span>,</span>
<span>                                     min_count <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co"># Calculating the library size factors</span></span>
<span><span class="co"># 1. Estimate the log2 fold change in mrna</span></span>
<span><span class="co"># converting counts to integer mode</span></span>
<span><span class="co"># 2. Estimate the log2 fold change in rpf</span></span>
<span><span class="co"># converting counts to integer mode</span></span>
<span><span class="co"># 3. Estimate the difference between two log2 fold changes</span></span>
<span><span class="co"># 4. Estimate the log2 ratio in first condition</span></span>
<span><span class="co"># converting counts to integer mode</span></span>
<span><span class="co"># 5. Estimate the log2 ratio in second condition</span></span>
<span><span class="co"># converting counts to integer mode</span></span>
<span><span class="co"># 6. Estimate the difference between two log2 ratios</span></span>
<span><span class="co"># Number of the log2FC and log2R used in determining the final p-value</span></span>
<span><span class="co"># log2FC: 7530</span></span>
<span><span class="co"># log2R: 5484</span></span>
<span></span>
<span></span>
<span><span class="co"># check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">te.xtail</span><span class="op">)</span></span>
<span><span class="co"># List of 2</span></span>
<span><span class="co"># $ deg_raw :Formal class 'xtail' [package ""] with 4 slots</span></span>
<span><span class="co"># .. ..@ listData       :List of 5</span></span>
<span><span class="co"># .. .. ..$ resultsTable        :Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co"># .. .. .. .. ..@ rownames       : chr [1:13014] "A4GALT" "AAAS" "AACS" "AADAT" ...</span></span>
<span><span class="co"># .. .. .. .. ..@ nrows          : int 13014</span></span>
<span><span class="co"># .. .. .. .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co"># .. .. .. .. ..@ elementMetadata: NULL</span></span>
<span><span class="co"># .. .. .. .. ..@ metadata       : list()</span></span>
<span><span class="co"># .. .. .. .. ..@ listData       :List of 11</span></span>
<span><span class="co"># .. .. .. .. .. ..$ mRNA_log2FC    : num [1:13014] 0.907 -0.283 -0.172 -0.32 -0.189 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ RPF_log2FC     : num [1:13014] 0.908 -0.481 -0.134 -0.487 -0.145 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ log2FC_TE_v1   : num [1:13014] 0.00488 -0.19712 0.03861 -0.16493 0.04371 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ pvalue_v1      : num [1:13014] 0.938 0.41 0.861 0.632 0.813 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ control_log2TE : num [1:13014] -2.485 1.119 1.128 0.458 0.722 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ treated_log2TE : num [1:13014] -2.484 0.921 1.166 0.291 0.765 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ log2FC_TE_v2   : num [1:13014] 0.0106 -0.1984 0.0378 -0.167 0.0431 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ pvalue_v2      : num [1:13014] 0.985 0.309 0.823 0.618 0.773 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ log2FC_TE_final: num [1:13014] 0.0106 -0.1971 0.0386 -0.1649 0.0437 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ pvalue_final   : num [1:13014] 0.985 0.41 0.861 0.632 0.813 ...</span></span>
<span><span class="co"># .. .. .. .. .. ..$ pvalue.adjust  : num [1:13014] 1 1 1 1 1 ...</span></span>
<span><span class="co"># .. .. ..$ log2FC_determine_num: int 7530</span></span>
<span><span class="co"># .. .. ..$ log2R_determine_num : int 5484</span></span>
<span><span class="co"># .. .. ..$ condition1          : chr "control"</span></span>
<span><span class="co"># .. .. ..$ condition2          : chr "treated"</span></span>
<span><span class="co"># .. ..@ elementType    : chr "ANY"</span></span>
<span><span class="co"># .. ..@ elementMetadata: NULL</span></span>
<span><span class="co"># .. ..@ metadata       : list()</span></span>
<span><span class="co"># $ deg_anno:'data.frame':  13014 obs. of  10 variables:</span></span>
<span><span class="co">#   ..$ log2FC_TE_v1   : num [1:13014] 0.00488 -0.19712 0.03861 -0.16493 0.04371 ...</span></span>
<span><span class="co"># ..$ pvalue_v1      : num [1:13014] 0.938 0.41 0.861 0.632 0.813 ...</span></span>
<span><span class="co"># ..$ treated_log2TE : num [1:13014] -2.484 0.921 1.166 0.291 0.765 ...</span></span>
<span><span class="co"># ..$ log2FC_TE_v2   : num [1:13014] 0.0106 -0.1984 0.0378 -0.167 0.0431 ...</span></span>
<span><span class="co"># ..$ pvalue_v2      : num [1:13014] 0.985 0.309 0.823 0.618 0.773 ...</span></span>
<span><span class="co"># ..$ log2FC_TE_final: num [1:13014] 0.0106 -0.1971 0.0386 -0.1649 0.0437 ...</span></span>
<span><span class="co"># ..$ pvalue_final   : num [1:13014] 0.985 0.41 0.861 0.632 0.813 ...</span></span>
<span><span class="co"># ..$ pvalue.adjust  : num [1:13014] 1 1 1 1 1 ...</span></span>
<span><span class="co"># ..$ type           : chr [1:13014] "nonSig" "nonSig" "nonSig" "nonSig" ...</span></span>
<span><span class="co"># ..$ gene_name      : chr [1:13014] "A4GALT" "AAAS" "AACS" "AADAT" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>GitHub - <strong>smithlabcode/riborex: Riborex</strong>: Fast and flexible identification of differential translation from Ribo-seq data</p></li>
<li><p>GitHub - <strong>xryanglab/xtail</strong>: Genome-wide assessment of differential translations with ribosome profiling data</p></li>
</ul>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/github\.com\/junjunlab\/riboTransVis");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./Track-plot.html" class="pagination-link" aria-label="Track plot">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Track plot</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Pathway-enrichment-analysis.html" class="pagination-link" aria-label="Pathway enrichment analysis">
        <span class="nav-page-text">Pathway enrichment analysis</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>riboTransVis documentation was written by Jun Zhang.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/junjunlab/riboTransVis/edit/main/Gene-expression-analysis.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/junjunlab/riboTransVis/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>